File: backend/src/metrics/registry.ts
// 轻量 Prometheus 文本格式指标注册表（无外部依赖）
// 采集：HTTP 请求计数/耗时、进程基础信息、PostgreSQL 连接池状态
import { APP_CONSTANTS } from '../config/env.js';

export const HTTP_DURATION_BUCKETS = [0.01, 0.05, 0.1, 0.5, 1, 2.5, 10];

const requestCounts = new Map<string, number>();
const durationBuckets = new Map<string, number[]>();
const durationCounts = new Map<string, number>();
let durationSumMs = 0;
let durationCount = 0;

function keyOf(method: string, route: string): string {
    return `${method} ${route}`;
}

function ensureDuration(key: string): number[] {
    let buckets = durationBuckets.get(key);
    if (buckets === undefined) {
        buckets = new Array(HTTP_DURATION_BUCKETS.length).fill(0);
        durationBuckets.set(key, buckets);
    }
    return buckets;
}

export function recordHttpRequest(method: string, route: string, status: number, durationMs: number): void {
    const key = keyOf(method, route);
    requestCounts.set(`${key} ${status}`, (requestCounts.get(`${key} ${status}`) ?? 0) + 1);
    durationCounts.set(key, (durationCounts.get(key) ?? 0) + 1);

    const buckets = ensureDuration(key);
    const seconds = durationMs / 1000;
    for (let index = 0; index < buckets.length; index++) {
        const threshold = HTTP_DURATION_BUCKETS[index];
        if (threshold !== undefined && seconds <= threshold) {
            buckets[index] = (buckets[index] ?? 0) + 1;
        }
    }
    durationSumMs += durationMs;
    durationCount++;
}

function escapeLabelValue(value: string): string {
    return value.replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, '\\n');
}

function renderCounters(): string[] {
    const lines: string[] = [
        '# HELP http_requests_total 处理的 HTTP 请求总数（按方法/路由/状态码）',
        '# TYPE http_requests_total counter',
    ];
    for (const [key, count] of [...requestCounts.entries()].sort()) {
        const [method = '', route = '', status = ''] = key.split(' ');
        lines.push(
            `http_requests_total{method="${escapeLabelValue(method)}",route="${escapeLabelValue(route)}",status="${status}"} ${count}`
        );
    }
    return lines;
}

function renderDurations(): string[] {
    const lines: string[] = [
        '# HELP http_request_duration_seconds HTTP 请求耗时分布（秒）',
        '# TYPE http_request_duration_seconds histogram',
    ];
    for (const [key, buckets] of [...durationBuckets.entries()].sort()) {
        const [method = '', route = ''] = key.split(' ');
        const labelPrefix = `method="${escapeLabelValue(method)}",route="${escapeLabelValue(route)}"`;
        buckets.forEach((count, index) => {
            lines.push(
                `http_request_duration_seconds_bucket{${labelPrefix},le="${HTTP_DURATION_BUCKETS[index]}"} ${count}`
            );
        });
        lines.push(`http_request_duration_seconds_bucket{${labelPrefix},le="+Inf"} ${durationCounts.get(key) ?? 0}`);
    }
    lines.push(`http_request_duration_seconds_sum ${durationSumMs / 1000}`);
    lines.push(`http_request_duration_seconds_count ${durationCount}`);
    return lines;
}

export interface PoolSnapshot {
    total: number;
    idle: number;
    waiting: number;
}

// 采集时由调用方注入连接池快照，指标注册表不直接依赖 pg
export function renderMetrics(snapshot: PoolSnapshot): string {
    const lines: string[] = [
        `# HELP process_uptime_seconds 进程运行时长（秒）`,
        `# TYPE process_uptime_seconds gauge`,
        `process_uptime_seconds ${process.uptime()}`,
        `# HELP process_memory_heap_bytes 进程堆内存（字节）`,
        `# TYPE process_memory_heap_bytes gauge`,
        `process_memory_heap_bytes ${process.memoryUsage().heapUsed}`,
        `# HELP pg_pool_total PostgreSQL 连接池总连接数`,
        `# TYPE pg_pool_total gauge`,
        `pg_pool_total ${snapshot.total}`,
        `# HELP pg_pool_idle PostgreSQL 连接池空闲连接数`,
        `# TYPE pg_pool_idle gauge`,
        `pg_pool_idle ${snapshot.idle}`,
        `# HELP pg_pool_waiting PostgreSQL 连接池等待连接数`,
        `# TYPE pg_pool_waiting gauge`,
        `pg_pool_waiting ${snapshot.waiting}`,
        `# HELP backend_info 后端版本信息`,
        `# TYPE backend_info gauge`,
        `backend_info{version="${APP_CONSTANTS.SERVICE_VERSION}"} 1`,
        ...renderCounters(),
        ...renderDurations(),
    ];
    return `${lines.join('\n')}\n`;
}

File: backend/internal/metrics/registry.go
package metrics

import (
	"fmt"
	"runtime"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

// HTTPDurationBuckets matches the previous backend's histogram boundaries.
var HTTPDurationBuckets = []float64{0.01, 0.05, 0.1, 0.5, 1, 2.5, 10}

// Registry is an in-memory Prometheus text-format store mirroring the
// previous backend's metric names and label shapes.
type Registry struct {
	mu              sync.Mutex
	startedAt       time.Time
	version         string
	counters        map[string]int64 // key: "method route status"
	durationBuckets map[string][]int64
	durationCounts  map[string]int64
	durationSumMS   float64
	durationTotal   int64
}

// NewRegistry creates a registry with the given histogram bucket boundaries.
func NewRegistry(version string, buckets []float64) *Registry {
	if len(buckets) == 0 {
		buckets = HTTPDurationBuckets
	}
	return &Registry{
		startedAt:       time.Now(),
		version:         version,
		counters:        make(map[string]int64),
		durationBuckets: make(map[string][]int64),
		durationCounts:  make(map[string]int64),
	}
}

// RecordRequest counts one request and observes its duration.
func (r *Registry) RecordRequest(method, route string, status int, durationMs float64) {
	r.mu.Lock()
	defer r.mu.Unlock()

	countKey := method + " " + route + " " + strconv.Itoa(status)
	r.counters[countKey]++

	histKey := method + " " + route
	r.durationCounts[histKey]++
	r.durationSumMS += durationMs
	r.durationTotal++

	buckets := r.durationBuckets[histKey]
	if buckets == nil {
		buckets = make([]int64, len(HTTPDurationBuckets))
	}
	seconds := durationMs / 1000
	for i, threshold := range HTTPDurationBuckets {
		if seconds <= threshold {
			buckets[i]++
		}
	}
	r.durationBuckets[histKey] = buckets
}

func escapeLabelValue(value string) string {
	value = strings.ReplaceAll(value, `\`, `\\`)
	value = strings.ReplaceAll(value, `"`, `\"`)
	return strings.ReplaceAll(value, "\n", `\n`)
}

// Render produces the Prometheus text exposition format (version 0.0.4).
func (r *Registry) Render(poolTotal, poolIdle, poolWaiting int64) string {
	r.mu.Lock()
	defer r.mu.Unlock()

	var sb strings.Builder
	var mem runtime.MemStats
	runtime.ReadMemStats(&mem)

	sb.WriteString("# HELP process_uptime_seconds 进程运行时长（秒）\n")
	sb.WriteString("# TYPE process_uptime_seconds gauge\n")
	fmt.Fprintf(&sb, "process_uptime_seconds %.0f\n", time.Since(r.startedAt).Seconds())

	sb.WriteString("# HELP process_memory_heap_bytes 进程堆内存（字节）\n")
	sb.WriteString("# TYPE process_memory_heap_bytes gauge\n")
	fmt.Fprintf(&sb, "process_memory_heap_bytes %d\n", mem.HeapAlloc)

	sb.WriteString("# HELP pg_pool_total PostgreSQL 连接池总连接数\n")
	sb.WriteString("# TYPE pg_pool_total gauge\n")
	fmt.Fprintf(&sb, "pg_pool_total %d\n", poolTotal)
	sb.WriteString("# HELP pg_pool_idle PostgreSQL 连接池空闲连接数\n")
	sb.WriteString("# TYPE pg_pool_idle gauge\n")
	fmt.Fprintf(&sb, "pg_pool_idle %d\n", poolIdle)
	sb.WriteString("# HELP pg_pool_waiting PostgreSQL 连接池等待连接数\n")
	sb.WriteString("# TYPE pg_pool_waiting gauge\n")
	fmt.Fprintf(&sb, "pg_pool_waiting %d\n", poolWaiting)

	sb.WriteString("# HELP backend_info 后端版本信息\n")
	sb.WriteString("# TYPE backend_info gauge\n")
	fmt.Fprintf(&sb, "backend_info{version=%q} 1\n", r.version)

	sb.WriteString("# HELP http_requests_total 处理的 HTTP 请求总数（按方法/路由/状态码）\n")
	sb.WriteString("# TYPE http_requests_total counter\n")
	countKeys := make([]string, 0, len(r.counters))
	for key := range r.counters {
		countKeys = append(countKeys, key)
	}
	sort.Strings(countKeys)
	for _, key := range countKeys {
		parts := strings.SplitN(key, " ", 3)
		method, route, status := parts[0], parts[1], parts[2]
		fmt.Fprintf(&sb, "http_requests_total{method=%q,route=%q,status=%q} %d\n",
			escapeLabelValue(method), escapeLabelValue(route), status, r.counters[key])
	}

	sb.WriteString("# HELP http_request_duration_seconds HTTP 请求耗时分布（秒）\n")
	sb.WriteString("# TYPE http_request_duration_seconds histogram\n")
	histKeys := make([]string, 0, len(r.durationBuckets))
	for key := range r.durationBuckets {
		histKeys = append(histKeys, key)
	}
	sort.Strings(histKeys)
	for _, key := range histKeys {
		parts := strings.SplitN(key, " ", 2)
		method, route := parts[0], parts[1]
		labelPrefix := fmt.Sprintf("method=%q,route=%q", escapeLabelValue(method), escapeLabelValue(route))
		for i, count := range r.durationBuckets[key] {
			fmt.Fprintf(&sb, "http_request_duration_seconds_bucket{%s,le=%q} %d\n",
				labelPrefix, fmt.Sprint(HTTPDurationBuckets[i]), count)
		}
		fmt.Fprintf(&sb, "http_request_duration_seconds_bucket{%s,le=%q} %d\n",
			labelPrefix, "+Inf", r.durationCounts[key])
	}
	fmt.Fprintf(&sb, "http_request_duration_seconds_sum %.6f\n", r.durationSumMS/1000)
	fmt.Fprintf(&sb, "http_request_duration_seconds_count %d\n", r.durationTotal)

	return sb.String()
}
